Wrap left-over builtins into DS class for categorization and autocomplete

aparajit-pratap opened the pull request:
### Purpose

This PR addresses https://jira.autodesk.com/browse/DYN-648 for wrapping left-over builtins from #7612 into a DS class so that they can be grouped together with the rest of the `List` class methods and be invoked from autocomplete using "List.". 

A new "BuiltIns.ds" file is being imported into the VM that contains a DS class called `List` deriving from the `DSCore.List` FFI class. This class wraps calls to the remaining DS implemented built-in methods that could not be ported over to ZT nodes. As a result these builtin methods can now be invoked in CBN's as member functions of the `List` class together with the existing methods on the base `DSCore.List` class. 

The existing Builtin methods are made internal so that they can no longer be invoked globally.

### Declarations

- [X] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [X] Snapshot of UI changes, if any.
![image](https://cloud.githubusercontent.com/assets/5710686/25374445/09b91dec-29d0-11e7-8ffe-314a20cd7f37.png)

- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@sharadkjaiswal 

### FYIs

@Benglin @ke-yu

sharadkjaiswal commented on src/Engine/ProtoCore/ExtendedLibraries/BuiltIns.ds:
```diff
@@ -0,0 +1,64 @@
+class List extends DSCore.List
```
shouldn't this file `import("DSCore.dll")` ?

aparajit-pratap replied:
That is already imported in the sequence of importing preloaded libraries but we can have the import statement here just in case the sequence is not followed.

sharadkjaiswal commented on src/Engine/ProtoCore/ClassTable.cs:
```diff
@@ -42,6 +43,27 @@ public bool IsEmpty
             }
         }
 
+        public bool IsFfiDerivedDsClass
```
I am wondering if we really need this special property. What if we say that a DS class can be imported from FFI class only if the FFI class is marked static or all its method and properties are static? So we may think of having a property called `IsStaticClass`.

aparajit-pratap replied:
So you're proposing implementing a concept of static class in DS?

sharadkjaiswal commented on src/Engine/ProtoCore/Reflection/Mirror.cs:
```diff
@@ -157,7 +158,8 @@ public static IEnumerable<ClassMirror> GetAllTypes(Core core)
                     Where(x => !CoreUtils.StartsWithSingleUnderscore(x.Name)
-                    && x.Name != DSDefinitions.Keyword.PointerReserved).
+                    && x.Name != DSDefinitions.Keyword.PointerReserved
+                    && !x.HasDSDerivedClass).
```
So we are not returning class nodes from FFI if it has derived class? I don't think we need to do this, you can use the attribute to hide the base class in library and autocompletion.

aparajit-pratap replied:
If I hide the base class using `IsVisibleInLibrary(false)` autocomplete will hide the members too.

aparajit-pratap commented:
@sharadkjaiswal I have addressed the review comments as discussed, please take a look. 
- Added checks to allow only derivation of static FFI classes
- Removed `IsFFIDerivedDSClass` and `HasDSDerivedCLass` properties

Thanks!

Randy-Ma commented on src/Engine/ProtoCore/FFI/CLRDLLModule.cs:
```diff
@@ -1294,6 +1297,10 @@ public FFIMethodAttributes(MethodInfo method, Dictionary<MethodInfo, Attribute[]
                 {
                     HiddenInLibrary = true;
                 }
+                else if (attr is IsVisibleInDynamoLibraryAttribute)
```
attr.Visible can be false.
